tensorflow/core/framework/memory_types.cc

```diff
@@ -104,23 +107,44 @@ Status MemoryTypesForNode(const OpRegistryInterface* op_registry,
     return it != ndef.attr().end() && it->second.b();
   }();
 
-  bool has_kernel_def = status.ok() && !IsFunctionCallOp(ndef.op());
+  bool is_fn = IsFunctionCallOp(ndef.op());
+  bool has_kernel_def = status.ok() && !is_fn;
   auto host_memory_required = [&](const DataType& dt) {
-    bool int32_on_device =
+    bool int32_on_device = 
         has_kernel_def || device_type.type_string() == "TPU" || has_xla_compile;
     return DataTypeAlwaysOnHost(dt) || (dt == DT_INT32 && !int32_on_device);
   };
-
-  if (has_kernel_def) {
+  
+  //  Edge cases:
+  //  1. If[Tcond=DT_BOOL, Tin=[DT_FLOAT,DT_INT32], Tout=[DT_FLOAT,DT_INT32]]
+  //     * Tcond marked HostMemory by kernel_def
+  //     * Tin[1] should be on host since it's INT32
+  //  2. ResourceGather[Tindices=DT_INT32](resource, indices)
+  //     * 'resource' marked HostMemory on the host
+  //     *  'indices' must be on the device (because the kernel expects it so)
+  //       - therefore, ignore the type
+
+  // Set all datatype to DEVICE_MEMORY by default, later on change it to
+  // HOST_MEMORY where it is required by the datatype.
+
+  if (has_kernel_def || is_fn) {
     // Gets the input/output names and their corresponding endpoint ranges.
     NameRangeMap inp_names;
     NameRangeMap out_names;
     TF_RETURN_IF_ERROR(
         NameRangesForNode(ndef, *op_def, &inp_names, &out_names));
 
     // Now that we know the size, fill with the default 'DEVICE_MEMORY'.
-    inp_mtypes->resize(GetTotal(inp_names), DEVICE_MEMORY);
-    out_mtypes->resize(GetTotal(out_names), DEVICE_MEMORY);
+    // Stick to old behavior except for function calls
+    // (since some kernels incorrectly rely on INT32 being on the GPU)
+    if(is_fn) {
+      inp_mtypes->resize(inp_dtypes.size(), DEVICE_MEMORY);
+      out_mtypes->resize(out_dtypes.size(), DEVICE_MEMORY);
+    }
+    else {
+      inp_mtypes->resize(GetTotal(inp_names), DEVICE_MEMORY);
+      out_mtypes->resize(GetTotal(out_names), DEVICE_MEMORY);
+    }
 
     // Fills in host memory types based on the kernel def.
     const auto& from_proto = kdef->host_memory_arg();
```

Comment:
Think this may need to be guarded with `if (has_kernel_def)`. We can get here for a node that has no kernel_def,  but is a function_call_op (assuming such a node is possible)

Reply:
I don't see how it's possible for an op to have no kernel_def, and my tests don't turn up any such cases, but yes, it's best to be safe. Although not with if(has_kernel_def). Note that the variable is poorly named (the name is retained from the original code). It should be guarded with if(kdef!=0).